Source/Streamstone.Tests/Scenarios/Tracking_entity_changes.cs

```diff
@@ -460,6 +470,113 @@ public void When_Null_followed_by_Delete()
         }
 
+        /********* Insert-Or-Merge or Insert-Or-Replace followed by XXX ************/
+        [TestCaseSource(nameof(GetThrowingOperationsForInsertOrMergeOrReplace))]
+        public void ThrowOnPrecedingInsertOrMergeOrReplaceWithAnything(Include first, Include second)
+        {
+            var events = new[]
+            {
+                CreateEvent(first),
+                CreateEvent(second),
+            };
+
+            Assert.Throws<InvalidOperationException>(() =>
+                    Stream.Write(stream, events));
+        }
+
+        [Test]
+        public void When_Null_followed_by_Insert_Or_Merge()
+        {
+            var entity = new TestEntity(EntityRowKey);
+
+            var events = new[]
+            {
+                CreateEvent(Include.Insert(entity)), // that combination
+                CreateEvent(Include.Delete(entity)), //  produces NULL
+                
+                CreateEvent(Include.InsertOrMerge(entity))
+            };
+
+            Stream.Write(stream, events);
+
+            var stored = RetrieveTestEntity(EntityRowKey);
+            Assert.That(stored, Is.Not.Null);
+        }
+
+        [Test]
+        public void When_Null_followed_by_Insert_Or_Replace()
+        {
+            var entity = new TestEntity(EntityRowKey);
+
+            var events = new[]
+            {
+                CreateEvent(Include.Insert(entity)), // that combination
+                CreateEvent(Include.Delete(entity)), //  produces NULL
+                
+                CreateEvent(Include.InsertOrMerge(entity))
+            };
+
+            Stream.Write(stream, events);
+
+            var stored = RetrieveTestEntity(EntityRowKey);
+            Assert.That(stored, Is.Not.Null);
+        }
+
+        [Test]
+        public void When_Insert_Or_Replace_followed_by_Insert_Or_Replace()
+        {
+            var entity = new TestEntity(EntityRowKey);
+            InsertTestEntity(entity);
+
+            entity.Data = "zzz";
+            var events = new[]
+            {
+                CreateEvent(Include.InsertOrReplace(entity)),
+                CreateEvent(Include.InsertOrReplace(entity))
+            };
+
+            Stream.Write(stream, events);
+
+            var stored = RetrieveTestEntity(EntityRowKey);
+            Assert.That(stored.Data, Is.EqualTo("zzz"));
+        }
+
+        [Test]
+        public void When_Insert_Or_Merge_followed_by_Insert_Or_Merge()
+        {
+            var entity = new TestEntity(EntityRowKey);
+            InsertTestEntity(entity);
+
+            entity.Data = "zzz";
+            var events = new[]
+            {
+                CreateEvent(Include.InsertOrMerge(entity)),
+                CreateEvent(Include.InsertOrMerge(entity))
+            };
+
+            Stream.Write(stream, events);
+
+            var stored = RetrieveTestEntity(EntityRowKey);
+            Assert.That(stored.Data, Is.EqualTo("zzz"));
```

Comment:
I think that actual merge is not covered by this test. Can you show that entity properties were really merged? Maybe use another entity with second event with a different  set of fields to showcase in-memory merge operation?